feat(compiler): support hexadecimal/octal/binary number literals

trotyl opened the pull request:
closes #12716

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #12716


## What is the new behavior?

Hexadecimal, octal, and binary literals allowed in template:

```html
<div>{{ 100 + 0x50 + 0o5 + 0b1 }}</div>
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

`012` is not supposed to be supported due to all templates are still module code, which always have `strict mode` turned on, making that being a syntax error. But current implementation treat it as decimal `12` which is incorrect. Now throw error for it.

Given that `012 === 12` is already the wrong result (should be `012 === 10` in non-strict mode), forbid that would not be consider a breaking change.

## Other information

The current validation for float and exponent number are not strict enough, invalid usage like `1e2e3` and `1.2.3` are allowed. Didn't change that.

trotyl commented:
Unknown fs flake in unit tests.

trotyl commented on packages/compiler/src/expression_parser/lexer.ts:
```diff
@@ -383,7 +417,21 @@ function unescape(code: number): number {
 
 function parseIntAutoRadix(text: string): number {
-  const result: number = parseInt(text);
+  const possibleRadixCharCode = text.charCodeAt(1);
```
This can be simplified by delegating to `Number()` constructor function, but would require [additional polyfills](https://github.com/zloirock/core-js/blob/678ff645c824cceb3475510c1ced043e4010e24a/modules/es6.number.constructor.js#L46-L68) at runtime. The problem is when polyfill not included it would not throw but just got `NaN` instead, not good for debugging.

<sub>Note that both `parseInt` and `Number.parseInt` do not support radix prefix.</sub>

vicb commented on packages/compiler/src/expression_parser/lexer.ts:
```diff
@@ -383,7 +417,21 @@ function unescape(code: number): number {
-  const result: number = parseInt(text);
+  const possibleRadixCharCode = text.charCodeAt(1);
+  let radix = 10;
+  if (isHexadecimalSeparator(possibleRadixCharCode)) {
```
should this test for `0x` instead ? (IIUC `\dx` would be detected as hex ?)

trotyl replied:
This function only applies to number, and there is a check for asserting first value is 0 above, so `1x12` is not possible to be got here.

vicb commented on packages/compiler/src/expression_parser/lexer.ts:
```diff
@@ -361,6 +393,34 @@ function isExponentSign(code: number): boolean {
+}
+
+function isDigitInRange(code: number, radix: number | null): boolean {
+  if (radix == null) {
```
`switch (radix)`

vicb commented on packages/compiler/src/expression_parser/lexer.ts:
```diff
@@ -382,8 +442,11 @@ function unescape(code: number): number {
 
-function parseIntAutoRadix(text: string): number {
-  const result: number = parseInt(text);
+function parseIntWithRadix(text: string, radix: number | null): number {
```
add a comment

vicb commented on packages/compiler/test/expression_parser/lexer_spec.ts:
```diff
@@ -94,12 +94,60 @@ function expectErrorToken(token: Token, index: any, message: string) {
+      it('should throw error for deprecated legacy octal', () => {
+        expectErrorToken(
+            lex('01234')[0], 0,
+            `Lexer Error: Legacy octal literals are not allowed in strict mode at column 0 in expression [01234]`);
```
seems like a breaking change, right ?

trotyl replied:
I'd say it's a better error message for the existing cryptic error, the previous behavior for parsing octals as decimals cannot be considering "working". But I can leave that intact if you insist.

trotyl committed:
feat(compiler): support hexadecimal/octal/binary number literals

Add support for hexadecimal (`0x123`)/octal(`0o123`)/binary(`0b10`)
number literals.

closes #12716

mhevery commented:
[presubmit](http://test/OCL:345516272:BASE:345516310:1607029111381:8e18a6fe)

mhevery commented:
Discussed with team and our default position is that we don't want to add new features to template syntax unless it provides sufficient value. While this request is reasonable, the value seems marginal, and does not seem to be worth the extra complexity documentation and mental model associated with it.
